### DISCUSSION TITLE
How to suppress events when async?

### DISCUSSION THREAD

**Discussion by wrekklol** [TolikPylypchuk/SharpHook#96] [Q&A] [answered]
I'm trying to figure out how to suppress events when running the hook async, but with no luck. If it isn't possible, how would I do it without async, but still not have it block my applications main thread?

**Comment by TolikPylypchuk** [comment, marked as answer]
Hello! Thanks for asking!

In order to have the ability to suppress events you need to use `SimpleGlobalHook` - `TaskPoolGlobalHook` doesn't support suppressing events since it runs event handlers on other threads.

It doesn't matter if you use `Run` or `RunAsync` to run the global hook - `SimpleGlobalHook` allows suppressing events either way.

Also, please note that suppressing events works only on Windows and macOS, it doesn't work on Linux.

**Reply by wrekklol** [reply]
Thank you, that seems to work. Though it is sadly a bit laggy now for me, as the handlers run synchronously.

Is there any way to make TaskPoolGlobalHook suppress events? Or if you have any other ideas I'd be very happy to hear them, as I think the library is amazingly simple to use. :)

**Reply by TolikPylypchuk** [reply]
Unfortunately, no, there is no way to suppress events with `TaskPoolGlobalHook` as suppression must be done on the same thread on which the event handler itself runs. The event handler inside `TaskPoolGlobalHook` delegates processing to other threads while the one in `SimpleGlobalHook` does not.

Maybe, you can push the decision to suppress the event or not as early as possible, suppress it (or not) and then continue processing the event on a different thread? I think it would be fairly easy with SharpHook.Reactive - if you're familiar with Rx.NET.

If you need to handle only keyboard events, you can create a keyboard-only global hook by passing `GlobalHookType.Keyboard` to the hook's constructor. This way at least the mouse won't lag if processing takes a long time.

Also, bear in mind that Windows can silently disable global hooks which take too long (without notifications of any kind).

Thanks for the kind words about the library, I really appreciate it!

**Reply by wrekklol** [reply]
I've never used Rx.NET, but I'll give it a go, thanks for your help! :)

**Reply by TolikPylypchuk** [reply]
Ah, well, if you've never used Rx.NET then such a thing won't really be easy :) There's a lot of stuff to learn and the learning curve is a little steep.

If you want to try anyway then you need to look into schedulers - they can be used to move event processing to other contexts.

**Reply by wrekklol** [reply]
I just got a thought, I know you cannot make two hooks currently, but what if you could? So you could have one for suppressing on the main thread, and then the one doing the events on another hook which runs on another thread? Sorry if that is a bit confusing.

**Reply by TolikPylypchuk** [reply]
That sounds a little too convoluted :)

You can always have more than one event handler on a single hook though.
